src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Telemetry/NavigatedTelemetryEvent.cs

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using NuGet.Common;
+using ContractsItemFilter = NuGet.VisualStudio.Internal.Contracts.ItemFilter;
+
+namespace NuGet.PackageManagement.Telemetry
+{
+    public class NavigatedTelemetryEvent : TelemetryEvent
+    {
+        internal const string NavigatedEventName = "Navigated";
+
+        internal const string NavigationTypePropertyName = "NavigationType";
+        internal const string OriginPropertyName = "Origin";
+        internal const string HyperLinkTypePropertyName = "HyperlinkType";
+
+        internal const string CurrentTabPropertyName = "CurrentTab";
+        internal const string IsSolutionViewPropertyName = "IsSolutionView";
+
+        internal const string PackageSourceMappingStatusPropertyName = "PackageSourceMappingStatus";
+
+        internal const string SourcesCountPropertyName = "SourcesCount";
+        internal const string IsGlobbingPropertyName = "IsGlobbing";
+
+        internal const string AlternativePackageIdPropertyName = "AlternativePackageId";
+
+        /// <summary>
+        /// General Navigation event with an origin specified.
+        /// </summary>
+        /// <param name="navigationType">Originating navigation type</param>
+        /// <param name="navigationOrigin">Control which started this navigation</param>
+        public NavigatedTelemetryEvent(NavigationType navigationType, NavigationOrigin navigationOrigin)
+            : base(NavigatedEventName)
+        {
+            base[NavigationTypePropertyName] = navigationType;
+            base[OriginPropertyName] = navigationOrigin;
+        }
+
+        /// <summary>
+        /// When adding PackageSourceMapping, includes additional data about the new mapping.
+        /// </summary>
+        public static NavigatedTelemetryEvent CreateWithAddPackageSourceMapping(int sourcesCount, bool isGlobbing)
```

Comment:
Rather than create `N` classes, or add properties directly from the product code, I decided on factory methods in the telemetry event itself. 
The idea is the event can be created with simply a navigation type/origin and appending properties is done in one place.